test/widgets/test_widgetbox.py

```diff
@@ -0,0 +1,78 @@
+import pytest
+
+from libqtile.backend.x11 import xcbq
+from libqtile.bar import Bar
+from libqtile.widget import TextBox, WidgetBox
+from test.conftest import BareConfig
+
+
+def no_op(*args, **kwargs):
+    pass
+
+
+class FakeWindow:
+    class _NestedWindow:
+        wid = 10
+
+    window = _NestedWindow()
+
+
+widget_config = pytest.mark.parametrize("qtile", [BareConfig], indirect=True)
+
+
+@widget_config
+def test_widgetbox_widget(qtile):
+    qtile.conn = xcbq.Connection(qtile.display)
+    qtile.test_xeyes()
+    qtile.windows_map = {}
+
+    # Disconnect functions called by widget but aren't needed here
+    qtile.call_soon = no_op
+    qtile.register_widget = no_op
```

Comment:
Perhaps confusingly, the `qtile` object here isn't actually `libqtile.qtile`, it's an instance of the `Qtile` class defined within "test/conftest.py". With this we interact with the running qtile process using a command client at `qtile.c`. In any case, leaving functions like these intact should be fine (indeed, that is probably preferable so that the test objects are as close as possible to the normal ones).

Reply:
Maybe the comment about "disconnecting" the functions is inappropriate. Without those lines I get an error `AttributeError: 'Qtile' object has no attribute 'call_soon'`. Your explanation that it's a different object would seem to explain why that happens.

I take your point about tests being as close as possible to normal operation but I'm not very familiar with pytest or how you can use qtile objects outside of qtile itself.

Reply:
No worries at all, tests do take a bit of getting used to.

OK that makes sense then. Maybe just update the comment so it's clear that we're tricking the widget into thinking `qtile` is `libqtile.qtile`. The rest of the testing looks good!